data_steward/cdr_cleaner/cleaning_rules/null_invalid_foreign_keys.py

```diff
@@ -21,84 +29,173 @@
+class NullInvalidForeignKeys(BaseCleaningRule):
     """
-    return '_mapping_' + domain_table
-
-
-def null_invalid_foreign_keys(project_id, dataset_id, sandbox_dataset_id=None):
+    Ensure invalid foreign keys are null while the remainder of the rows persist
     """
-    This method gets the queries required to make invalid foreign keys null
 
-    :param project_id: Project associated with the input and output datasets
-    :param dataset_id: Dataset where cleaning rules are to be applied
-    :param sandbox_dataset_id: Identifies the sandbox dataset to store rows 
-    #TODO use sandbox_dataset_id for CR
-    :return: a list of queries
-    """
-    queries_list = []
-    for table in resources.CDM_TABLES:
-        field_names = [field['name'] for field in resources.fields_for(table)]
-        foreign_keys_flags = []
-        fields_to_join = []
-
-        for field_name in field_names:
-            if field_name in FOREIGN_KEYS_FIELDS and field_name != table + '_id':
-                fields_to_join.append(field_name)
-                foreign_keys_flags.append(field_name)
-
-        if fields_to_join:
-            col_exprs = []
-            for field in field_names:
-                if field in fields_to_join:
-                    if field in foreign_keys_flags:
-                        col_expr = '{x}.'.format(x=field[:3]) + field
-                else:
-                    col_expr = field
-                col_exprs.append(col_expr)
-            cols = ', '.join(col_exprs)
-
-            join_expression = []
-            for key in FOREIGN_KEYS_FIELDS:
-                if key in foreign_keys_flags:
-                    if key == 'person_id':
-                        table_alias = cdr_consts.PERSON_TABLE_NAME
+    def __init__(self, project_id, dataset_id, sandbox_dataset_id):
+        """
+        Initialize the class with proper information.
+
+        Set the issue numbers, description and affected datasets. As other tickets may affect this SQL,
+        append them to the list of Jira Issues.
+        DO NOT REMOVE ORIGINAL JIRA ISSUE NUMBERS!
+        """
+        desc = (
+            'Ensures that invalid foreign keys are null while the remainder of the rows persist'
+        )
+        super().__init__(issue_numbers=['DC-388', 'DC-807'],
+                         description=desc,
+                         affected_datasets=[cdr_consts.COMBINED],
+                         project_id=project_id,
+                         dataset_id=dataset_id,
+                         sandbox_dataset_id=sandbox_dataset_id,
+                         affected_tables=resources.CDM_TABLES)
+
+    def get_mapping_tables(self, domain_table):
+        """
+        Get name of mapping table generated for a domain table
+
+        :param domain_table: one of the domain tables (e.g. 'visit_occurrence',
+            'condition_occurrence')
+        :return: mapping table name
+        """
+        return '_mapping_' + domain_table
+
```

Comment:
I suggest breaking up the `get_query_specs` method and moving some parts of code to separate class methods because `get_query_specs` contains too much logic right now and it's hard to unit test this method. It would be much easier to test these individual methods in the unit test. And you could potentially mock those methods in `test_get_query_specs`. 

```suggestion
    def get_field_names(self, table):
        field_names = [field['name'] for field in resources.fields_for(table)]
        return field_names

    def get_foreign_keys(self, table):
        foreign_keys_flags = []
        for field_name in self.get_field_names(table):
            if field_name in FOREIGN_KEYS_FIELDS and field_name != table + '_id':
                foreign_keys_flags.append(field_name)
        return foreign_keys_flags

    def has_foreign_key(self, table):
        return len(self.get_foreign_keys(table)) > 0

    def get_col_expression(self, table):
        col_exprs = []
        foreign_keys = self.get_foreign_keys(table)
        for field in self.get_field_names(table):
            if field in foreign_keys:
                col_expr = '{x}.'.format(x=field[:3]) + field
            else:
                col_expr = field
            col_exprs.append(col_expr)
        return ', '.join(col_exprs)

    def get_join_expression(self, table):
        join_expression = []
        for key in FOREIGN_KEYS_FIELDS:
            if key in self.get_foreign_keys(table):
                if key == 'person_id':
                    table_alias = cdr_consts.PERSON_TABLE_NAME
                else:
                    table_alias = self.get_mapping_tables(
                        '{x}'.format(x=key)[:-3])
                join_expression.append(
                    LEFT_JOIN.render(dataset_id=self.dataset_id,
                                     prefix=key[:3],
                                     field=key,
                                     table=table_alias))
        return ' '.join(join_expression)
```